### DISCUSSION TITLE
Add channels to shop-api - Help needed

### DISCUSSION THREAD

**Discussion by AndersGerner** [vendurehq/vendure#1449] [Q&A] [answered]
Hi :) 

In my multi-tenant solution, I would like to extend the shop API with the "channels" query from the admin-api.
I have tried to read the current code and try to get an overview, but I can't seem to find the right way to do it.

I would assume the resolver would be something like:
```import { Args, Query, Resolver } from '@nestjs/graphql';
import { ChannelService, Ctx, RequestContext } from '@vendure/core'

@Resolver()
export class ChannelsResolver {

  constructor(private channelsService: ChannelService) {}

  @Query()
  channels(@Ctx() ctx: RequestContext) {
    return this.channelsService.findAll(ctx)
  }

}
```

And the plugin:
```
import gql from 'graphql-tag';
import { PluginCommonModule, VendurePlugin } from '@vendure/core';
import { GetChannelsService } from './plugins/channels/channels.service'
import { ChannelsResolver } from './plugins/channels/channels.resolver'

@VendurePlugin({
  imports: [PluginCommonModule],
  providers: [GetChannelsService],
  shopApiExtensions: {
    schema: gql`
      extend type Query {
        channels(): [Channel!]!
    }`,
    resolvers: [ChannelsResolver]
  }
})
export class ChannelsPlugin {}
```

But what should the service be?
```
import { Injectable } from '@nestjs/common';
import { Channel } from '@vendure/common/lib/generated-types';
import { RequestContext } from '@vendure/core';

@Injectable()
export class GetChannelsService {
    getChannels(ctx: RequestContext): Promise<Channel[]> { 
        return new Promise<Channel[]>((resolve, reject) => {
            
            
        })
    }
}
```

Thank you a lot in advance for any help given.

**Comment by michaelbromley** [comment, marked as answer]
Hi @AndersGerner,

If you are using the built-in `ChannelService` from `@vendure/core`, you don't need to implement your own service - it is already implemented. The plugin & resolver files alone should suffice. One fix you need to make is to remove the parentheses in your schema extension:
```diff
    schema: gql`
      extend type Query {
-        channels(): [Channel!]!
+        channels: [Channel!]!
    }`,
```
because the query does not take any arguments, so the presence of parentheses would cause a syntax error.

**Reply by AndersGerner** [reply]
Awesome @michaelbromley ! Both with the help to fix this but also the response time! That's simply amazing! This is a great first impression. 

Really looking forward to exploring vendure more - We are making a marketplace and found this to be the most interesting backend.

**Reply by michaelbromley** [reply]
Thanks for the great feedback! If you need more help, feel free to join our [Slack community](https://join.slack.com/t/vendure-ecommerce/shared_invite/enQtNzA1NTcyMDY3NTg0LTMzZGQzNDczOWJiMTU2YjAyNWJlMzdmZGE3ZDY5Y2RjMGYxZWNlYTI4NmU4Y2Q1MDNlYzE4MzQ5ODcyYTdmMGU) where you can usually get quick answers to issues you run into/questions you have as you explore the framework.
